File: Codeforces/Codeforces Contests/Contest 1058 Div 2/B_Distinct_Elements.py
def solve():
    t = int(input())
    for _ in range(t):
        n = int(input())
        b = list(map(int, input().split()))
        
        a = [0] * n
        
        for i in range(n):
            if i == 0:
                diff = b[0]
            else:
                diff = b[i] - b[i-1]
            
            last_pos = (i + 1) - diff
            
            if last_pos == 0:
                a[i] = i + 1
            else:
                a[i] = a[last_pos - 1]
        
        print(*a)

solve()

File: Codeforces/Codeforces Contests/Contest 1058 Div 2/B_Distinct_Elements.cpp
#include<bits/stdc++.h>
using namespace std;

#define int long long

void solve() {
    int n;
    cin >> n;

    vector<int> b(n);
    for(auto &x : b) {
        cin >> x;
    }

    vector<int> a(n);
    a[0] = 1;
    int num = 1;

    for(int i = 1; i < n; i++) {
        int index = i + 1;
        int diff = b[i] - b[i - 1];
        if(diff == index) {
            a[i] = ++num;
        }

        else {
            a[i] = a[i - diff];
        }
    }


    for(auto &x : a) cout << x << " ";
    cout << "\n";

}

signed main() {

    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--)
        solve();

    return 0;
}